### PR TITLE
Support security annotations for forms that dont need anti CSRF tokens.

### PR DESCRIPTION
**Pull request by psiinon**
Note that this includes a couple of changes that were implemented in #570 - I have updated my local copy of the CSRFCountermeasures tests but want to see it the original submitter was going to update them before adding my changes.

### CODE REVIEW

File: src/org/zaproxy/zap/extension/ascanrulesBeta/resources/Messages.properties [hunk 1]
```diff
@@ -23,6 +23,7 @@ ascanbeta.crossdomain.silverlight.soln = Configure the clientaccesspolicy.xml fi
 ascanbeta.crossdomain.silverlight.extrainfo = The web server permits malicious cross-domain requests originating from Silverlight components served from any third party domain, to this domain. If the victim user is logged into this service, the malicious requests are processed using the privileges of the victim, and can result in data from this service being compromised by an unauthorised third party web site, via the victim's web browsers. It can also result in Cross Site Request Forgery (CSRF) type attacks. This is particularly likely to be an issue if a Cookie based session implementation is in use.
 
 ascanbeta.csrftokenscan.name=Anti CSRF Tokens Scanner
+ascanbeta.csrftokenscan.extrainfo.annotation = This is an informational alert as the form has a security annotation indicating that is does not need an Anti CSRF Token. This should be tested manually to ensure the annotation is correct.
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/resources/Messages.properties]: I believe `Anti-CSRF` should be hyphenated.

File: src/org/zaproxy/zap/extension/ascanrulesBeta/ZapAddOn.xml [hunk 2]
```diff
@@ -9,6 +9,7 @@
 	<![CDATA[
 	Fix FP in "Source Code Disclosure - /WEB-INF folder" on successful responses (Issue 3048).<br>
 	Fix FP in "Integer Overflow Error" on 500 error responses (Issue 3064).<br>
+	Support security annotations for forms that dont need anti CSRF tokens.<br>
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/ZapAddOn.xml]: I believe `Anti-CSRF` should be hyphenated, and `don't` should have an apostrophe.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java [hunk 3]
```diff
@@ -209,22 +223,27 @@ public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
...
-			    Alert alert = new Alert(getPluginId(), Alert.RISK_LOW, Alert.CONFIDENCE_MEDIUM,  getName());
-			    alert.setDetail(
-			    			desc + "\n"+getDescription(), 
-				    		msg.getRequestHeader().getURI().toString(),
-				    		"",  //parameter: none.
-				    		"", 
-				    		extraInfo,
-				    		getSolution(), 
-				            getReference(), 
+				Alert alert = new Alert(getPluginId(), risk, Alert.CONFIDENCE_MEDIUM,  getName());
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Seems to be a bunch of formatting changes here.
https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting

File: src/org/zaproxy/zap/extension/pscanrulesBeta/ZapAddOn.xml [hunk 4]
```diff
@@ -1,17 +1,13 @@
...
 	<author>ZAP Dev Team</author>
 	<url/>
 	<changes>
 	<![CDATA[
-	Issue 2576: CSRFCountermeasures, remove unneeded Attack and Evidence messages.<br>
-	Issue 2574: CookieLooselyScopedScanner, remove attack field and update description.<br>
-	Support ignoring specified forms when checking for CSRF vulnerabilities.<br>
-	Correct the plugin ID of Absence of Anti-CSRF Tokens from 40014 to 10202.<br>
-	Issue 2860: Add further reference links.
+	Support security annotations for forms that dont need anti CSRF tokens.<br>
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/ZapAddOn.xml]: I believe `Anti-CSRF` should be hyphenated, and `don't` should have an apostrophe.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/resources/Messages.properties [hunk 5]
```diff
@@ -70,6 +70,7 @@ pscanbeta.insecurejsfviewstate.extrainfo=JSF ViewState [{0}] is insecure
 pscanbeta.noanticsrftokens.name=Absence of Anti-CSRF Tokens
 pscanbeta.noanticsrftokens.desc=No Anti-CSRF tokens were found in a HTML submission form.
 pscanbeta.noanticsrftokens.alert.extrainfo=No known Anti-CSRF tokens {0} were found in the following HTML forms: {1}.
+pscanbeta.noanticsrftokens.extrainfo.annotation=This is an informational alert as the form has a security annotation indicating that is does not need an Anti CSRF Token. This should be tested manually to ensure the annotation is correct.
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/resources/Messages.properties]: I believe `Anti-CSRF` should be hyphenated.

File: src/org/zaproxy/zap/extension/ascanrulesBeta/Csrftokenscan.java [hunk 6]
```diff
@@ -49,6 +51,8 @@
 	private static final int PLUGIN_ID = 20012;
 
 	private List<String> ignoreList = new ArrayList<String>();
+	private String ignoreAttName = null;
+	private String ignoreAttValue = null;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/Csrftokenscan.java]: The initialisations to null are not needed.

File: src/org/zaproxy/zap/extension/ascanrulesBeta/Csrftokenscan.java [hunk 7]
```diff
@@ -213,13 +219,24 @@ public void scan() {
...
-						bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, 
-								getBaseMsg().getRequestHeader().getURI().toString(),
-								"",	// No param 
-								"", // No attack
-								"", // No otherinfo
-								evidence,
-								getBaseMsg());
+
+						if (formHasSecurityAnnotation(formElement)) {
+							bingo(Alert.RISK_INFO, Alert.CONFIDENCE_MEDIUM,
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/Csrftokenscan.java]: Could use the same `bingo` in both cases (similar to the passive scanner).

File: src/org/zaproxy/zap/extension/ascanrulesBeta/Csrftokenscan.java [hunk 8]
```diff
@@ -249,9 +266,21 @@ private boolean formOnIgnoreList(Element formElement) {
 		return false;
 	}
 
-    private static boolean isHiddenInputElement(Element inputElement) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/Csrftokenscan.java]: Indentation could be kept the same.

File: src/org/zaproxy/zap/extension/ascanrulesBeta/resources/Messages.properties (hunk 1 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/resources/Messages.properties]: There's a typo(?) "that *it* does not".

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java [hunk 9]
```diff
@@ -293,4 +312,20 @@ public String getReference() {
...
+    protected ExtensionAntiCSRF getExtensionAntiCSRF() {
+        return (ExtensionAntiCSRF) Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.NAME);
+    }
+
+    protected String getCSRFIgnoreList() {
+        return Model.getSingleton().getOptionsParam().getConfig().getString("rules.csrf.ignorelist");
+    }
+
+    protected String getCSRFIgnoreAttName() {
+        return getConfig().getString("rules.csrf.ignore.attname", null);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Shouldn't this also use the main configuration? (The rule configurations are not being injected into the passive scanners, that's done only for active ones?)

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java [hunk 10]
```diff
@@ -293,4 +312,20 @@ public String getReference() {
...
+    }
+    
+    protected String getCSRFIgnoreAttValue() {
+        return getConfig().getString("rules.csrf.ignore.attvalue", null);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Ditto.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/resources/Messages.properties (hunk 5 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/resources/Messages.properties]: There's a typo(?) "that *it* does not".

**Comment by psiinon** [comment]
Hopefully fixed those issues

**Comment by psiinon** [comment]
@thc202 @kingthorin this ok now?

**Comment by thc202** [comment]
Thanks!

**Comment by kingthorin** [comment]
Yup :grinning:

**Comment by psiinon** [comment]
thanks :)

**Pull request merged**